### PR TITLE
Add mix debug mode. Close #4106.

### LINKED ISSUE #4106: Support MIX_DEBUG=1 that prints each task being run
We should add debug mode for Mix that will print each task being run.

* [ ] Introduce `Mix.debug?/0` (to read) and `Mix.debug/1` (to write)
* [ ] Mix.CLI should check if MIX_DEBUG is set and set the mode accordingly (default is false)
* [ ] If `Mix.debug?` is true, we should print whenever a task is invoked

Original report: #4099

### PR DESCRIPTION
**Pull request by tuvistavie**
I implemented mix debug option as described in #4106.

### CODE REVIEW

File: lib/mix/lib/mix/cli.ex [hunk 1]
```diff
@@ -9,6 +9,7 @@ defmodule Mix.CLI do
     Mix.Local.append_paths
 
     if System.get_env("MIX_QUIET"), do: Mix.shell(Mix.Shell.Quiet)
+    if System.get_env("MIX_DEBUG"), do: Mix.debug(true)
```
Comment (tuvistavie) [inline lib/mix/lib/mix/cli.ex]: Where should I document this? I found that `MIX_QUIET` is described in the man page, 
so it would be the right place to add info about this right?

**Comment by tuvistavie** [comment]
I would like to discuss a point. `MIX_QUIET` and `MIX_DEBUG` now only check if the variable is set,
which means that `MIX_DEBUG=0` and `MIX_DEBUG=1` behaves exactly the same way.
I find this behavior a little counter-intuitive, so maybe having a list of blacklist values, (no, false, 0 would be a good start)
that would avoid settings the flag even if the env variable is on would be good to have.
Any thoughts?

File: lib/mix/lib/mix/cli.ex (hunk 1 above)
Comment (josevalim) [inline lib/mix/lib/mix/cli.ex]: In the man page at least. We could probably add an environment variables section to the `Mix` module docs too. Could you please act on both? :)

**Comment by josevalim** [comment]
> that would avoid setting the flag even if the env variable is on would be good to have.

Let's check if it is "1" or "true".

File: lib/mix/lib/mix/task.ex [hunk 2]
```diff
@@ -275,6 +277,12 @@ defmodule Mix.Task do
     end
   end
 
+  defp output_task_debug_info(task, args) do
+    Mix.shell.info("Running 'mix #{task_to_string(task, args)}'")
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex]: Can we do: `"** Running mix #{task_to_string(task, args)} (inside #{inspect proj})"`. I don t think we need to wrap in quotes the mix command, it would be just noise in this case. :)

Comment (josevalim) [inline lib/mix/lib/mix/task.ex]: Btw, the `proj` may be nil, so we need to check accordingly. :)

File: lib/mix/test/mix/task_test.exs [hunk 3]
```diff
@@ -26,6 +26,17 @@ defmodule Mix.TaskTest do
...
 
+  test "output task debug info if Mix.debug? is true" do
+    Mix.shell Mix.Shell.IO
+    Mix.debug(true)
+
+    assert ExUnit.CaptureIO.capture_io(fn -> Mix.Task.run("hello") end) =~
+      "Running 'mix hello'"
+
+    Mix.shell(Mix.Shell.Process)
+    Mix.debug(false)
```
Comment (josevalim) [inline lib/mix/test/mix/task_test.exs]: Move those two last commands to an after clause:

```
after
  Mix.shell(Mix.Shell.Process)
  Mix.debug(false)
end
```

**Comment by josevalim** [comment]
Thank you! I have added some comments! :heart:

**Comment by tuvistavie** [comment]
@josevalim Thanks for the feedback!
i just updated my PR.
I split the doc update in `Mix` module to another commit as it was not specific to the `DEBUG` flag.

File: lib/mix/lib/mix.ex [hunk 4]
```diff
@@ -235,6 +252,20 @@ defmodule Mix do
...
+  Returns true if Mix is in debug mode.
+  """
+  def debug? do
+    Mix.State.get(:debug, false)
+  end
+
+  @doc """
+  Sets Mix debug mode.
+  """
+  def debug(debug) do
```
Comment (lexmag) [inline lib/mix/lib/mix.ex]: I think it worth to have `is_boolean/1` guard.

Comment (tuvistavie) [inline lib/mix/lib/mix.ex]: It does seem safer! I'll add this, thank you.

File: lib/mix/lib/mix/task.ex [hunk 5]
```diff
@@ -275,6 +277,13 @@ defmodule Mix.Task do
     end
   end
 
+  defp output_task_debug_info(task, args, proj) do
+    proj_str = if is_nil(proj), do: "", else: " (inside #{inspect proj})"
```
Comment (lexmag) [inline lib/mix/lib/mix/task.ex]: We could make an opposite order of `if` clauses:
```elixir
if proj, do: " (inside #{inspect proj})", else: ""
```

File: lib/mix/test/mix/cli_test.exs [hunk 6]
```diff
@@ -52,6 +52,15 @@ defmodule Mix.CLITest do
 
       assert contents =~ "Hello from MyProject!\n"
       refute contents =~ "This won't appear"
+
+      contents = mix ~w[my_hello], [{"MIX_QUIET", "0"}]
```
Comment (lexmag) [inline lib/mix/test/mix/cli_test.exs]: Very nitpicking: can we be consistent in `w` sigil definition and use `()`?

File: lib/mix/lib/mix/task.ex (hunk 5 above)
Comment (tuvistavie) [inline lib/mix/lib/mix/task.ex]: I usually stick with conditions that return booleans when possible but I agree

```elixir
if proj, do: " (inside #{inspect proj})", else: ""
```

read better.

File: lib/mix/test/mix/cli_test.exs (hunk 6 above)
Comment (tuvistavie) [inline lib/mix/test/mix/cli_test.exs]: All the sigil definitions in this file were using `~w[]` so I preferred to stick with it.
After checking, `cli_test.exs` and `test_helpers.exs` are the only files which use
`~w[]`, so we could fix this, but I think it would be better to make it as another PR,
as it is not really related to the original issue anymore.

Comment (lexmag) [inline lib/mix/test/mix/cli_test.exs]: Totally agree. :+1:

File: lib/mix/lib/mix.ex [hunk 7]
```diff
@@ -152,6 +152,23 @@ defmodule Mix do
   with extra clean up logic.
 
   Note aliases do not show up on `mix help`.
+
+  ## Environment variables
```
Comment (josevalim) [inline lib/mix/lib/mix.ex]: Sorry for the nitpick but please add new line after tiles. :)

File: lib/mix/lib/mix.ex [hunk 8]
```diff
@@ -152,6 +152,23 @@ defmodule Mix do
...
+  Environment variables can be used to modify Mix behaviour.
+
+  Mix responds to the following variables:
+
+  * `MIX_ARCHIVE` - allows specifying the directory into which the archives should be installed
```
Comment (josevalim) [inline lib/mix/lib/mix.ex]: Indent bullets with 2 spaces.

File: lib/mix/lib/mix/cli.ex [hunk 9]
```diff
@@ -71,6 +72,10 @@ defmodule Mix.CLI do
     end
   end
 
+  defp env_variable_activated?(name) do
```
Comment (josevalim) [inline lib/mix/lib/mix/cli.ex]: :heart:

File: lib/mix/lib/mix/task.ex [hunk 10]
```diff
@@ -275,6 +277,13 @@ defmodule Mix.Task do
     end
   end
 
+  defp output_task_debug_info(task, args, proj) do
+    proj_str = if proj, do: " (inside #{inspect proj})", else: ""
+    Mix.shell.info("** Running mix #{task_to_string(task, args)}" <> proj_str)
```
Comment (josevalim) [inline lib/mix/lib/mix/task.ex]: What about:

    "** Running mix " <> task_to_string(task, args) <> project_to_string(prog)

**Comment by josevalim** [comment]
@tuvistavie this is perfect. I have added one last batch of nitpicking commands and we are good to go. :)

File: lib/mix/lib/mix/task.ex (hunk 10 above)
Comment (lexmag) [inline lib/mix/lib/mix/task.ex]: Then we probably don't need `output_task_debug_info` function.

Comment (tuvistavie) [inline lib/mix/lib/mix/task.ex]: I moved `project_to_string` to its own function, but I would rather keep 
`output_task_debug_info` as a function to reduce the noise to `run_task/3`.

**Comment by tuvistavie** [comment]
@josevalim Thank you, I just pushed the last fixes.

File: lib/mix/lib/mix/task.ex [hunk 11]
```diff
@@ -275,6 +277,14 @@ defmodule Mix.Task do
     end
   end
 
+  defp output_task_debug_info(task, args, proj) do
+    Mix.shell.info("** Running mix " <> task_to_string(task, args) <> project_to_string(proj))
+  end
+  defp project_to_string(nil), do: ""
```
Comment (lexmag) [inline lib/mix/lib/mix/task.ex]: Very last comment, could you separate function definition with newlines. :bowtie:

Comment (tuvistavie) [inline lib/mix/lib/mix/task.ex]: Done! :+1:

Comment (lexmag) [inline lib/mix/lib/mix/task.ex]: :yellow_heart:

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
